File: 0279-perfect-squares/0279-perfect-squares.java
class Solution {
    
    HashMap<Integer,Integer> memo;
    int[] squareNums;
    public boolean isSquare(int n) {
        int x = (int) Math.sqrt(n);
        //System.out.println(x);
        return (x * x) == n;
    }
    public int dp(int n, int squareNumCount) {
        //System.out.println(n);
        if(n < 0) {
            return Integer.MAX_VALUE;
        }
        if(n == 0) return 0;
        if(memo.containsKey(n)) return memo.get(n);
        if(isSquare(n))  { 
            memo.put(n,1);
            return memo.get(n);
        }
        int ans = Integer.MAX_VALUE;
        for(int i = 1; i < squareNumCount; i++) {
            if(squareNums[i] <= n) {
                ans = Math.min(ans, 1 + dp(n - squareNums[i], squareNumCount));
            } 
        }
        memo.put(n,ans);
        return memo.get(n);
    }
    public int numSquares(int n) {
        //System.out.println(isSquare(n));
        memo = new HashMap<Integer, Integer>();
        int squareNumCount = (int) Math.sqrt(n) + 1;
        squareNums = new int[squareNumCount];
        for(int i = 1; i < squareNumCount; i++) {
            squareNums[i] = i * i;
        }
        return dp(n,squareNumCount);
    }
}

File: 0279-perfect-squares/0279-perfect-squares.js
/**
 * @param {number} n
 * @return {number}
 */
var numSquares = function(n) {
    
    let isSquare = (num) => {
        if(num >= 1) {
            let x = parseInt(Math.sqrt(num));
            return (x * x) == num;
        }
        return false;
    }
    
    let memo = new Map();
    
   let dp = (sum) => {
       if(sum < 0) {
           return Infinity; //not possible
       }
       if(sum  == 0) {
           return 0;
       }
       if(isSquare(sum)) {
           return 1;
       }
       if(memo.has(sum)) return memo.get(sum);
       let ans = Infinity;
       for(let i = sum; i >= 1; i--) {
           if(isSquare(i)) {
               ans = Math.min(ans, 1 + dp(sum - i));
           }
       }
       memo.set(sum, ans);
       return ans;
   }
   return dp(n);
    
};